plugins/producer_plugin/producer_plugin.cpp

```diff
@@ -1515,32 +1507,26 @@ producer_plugin_impl::start_block_result producer_plugin_impl::start_block() {
                int num_processed = 0;
                auto unapplied_trxs_size = _unapplied_transactions.size();
                auto calculate_transaction_category = [&](const transaction_metadata_ptr& trx) {
-                  if (trx->packed_trx()->expiration() < pending_block_time) {
-                     return tx_category::EXPIRED;
-                  } else if (persisted_by_id.find(trx->id()) != persisted_by_id.end()) {
+                  if (_unapplied_transactions.is_persisted(trx)) {
                      return tx_category::PERSISTED;
                   } else {
                      return tx_category::UNEXPIRED_UNPERSISTED;
                   }
                };
 
-               while( !exhausted ) {
-                  transaction_metadata_ptr trx = _unapplied_transactions.next();
-                  if( trx == nullptr ) break;
+               auto itr = _unapplied_transactions.begin();
+               while( itr != _unapplied_transactions.end() ) {
+                  if( preprocess_deadline <= fc::time_point::now() ) exhausted = true;
+                  if( exhausted ) break;
 
+                  const transaction_metadata_ptr trx = itr->trx_meta;
                   ++num_processed;
                   auto category = calculate_transaction_category(trx);
-                  if (category == tx_category::EXPIRED ||
-                     (category == tx_category::UNEXPIRED_UNPERSISTED && _producers.empty()))
+                  if (category == tx_category::UNEXPIRED_UNPERSISTED && _producers.empty())
                   {
                      ++num_failed;
-                     if (!_producers.empty()) {
-                        fc_dlog(_trx_trace_log, "[TRX_TRACE] Node with producers configured is dropping an EXPIRED transaction that was PREVIOUSLY ACCEPTED : ${txid}",
-                               ("txid", trx->id()));
-                     }
-                     if( preprocess_deadline <= fc::time_point::now() ) exhausted = true;
-                     continue; // _unapplied_transactions removed trx, so just continue
-
+                     itr = _unapplied_transactions.erase( itr );
+                     continue;
                   } else if (category == tx_category::PERSISTED ||
```

Comment:
This condition no longer needs to be checked given the above changes.

Reply:
Done